src/algebra/monoid_algebra/division.lean

```diff
@@ -0,0 +1,184 @@
+
+lemma div_of_add (x : add_monoid_algebra k G) (a b : G) :
+  div_of (a + b) x = div_of b (div_of a x) :=
+by { ext, simp only [add_monoid_algebra.div_of_apply, add_assoc] }
+
+lemma div_of_of'_mul (a : G) (x : add_monoid_algebra k G) :
+  div_of a (of' k G a * x) = x :=
+begin
+  ext b,
+  rw [add_monoid_algebra.div_of_apply, of'_apply, single_mul_apply_aux, one_mul],
+  intro c,
+  exact add_right_inj _,
+end
+
+lemma div_of_mul_of' (a : G) (x : add_monoid_algebra k G) :
+  div_of a (x * of' k G a) = x :=
+begin
+  ext b,
+  rw [add_monoid_algebra.div_of_apply, of'_apply, mul_single_apply_aux, mul_one],
+  intro c,
+  rw add_comm,
+  exact add_right_inj _,
+end
+
+lemma div_of_of' (a : G) : div_of a (of' k G a) = 1 :=
+by simpa only [one_mul] using div_of_mul_of' a (1 : add_monoid_algebra k G)
+
+/-- The remainder upon division by `of' k G g`. -/
+noncomputable def mod_of (g : G) (x : add_monoid_algebra k G) : add_monoid_algebra k G :=
+x.filter (λ g₁, ¬∃ g₂, g₁ = g + g₂)
+
+@[simp] lemma mod_of_apply_of_not_exists_add {g : G} (x : add_monoid_algebra k G) {g' : G}
+  (h : ¬∃ d, g' = g + d):
+  mod_of g x g' = x g' :=
+finsupp.filter_apply_pos _ _ h
+
+@[simp] lemma mod_of_apply_of_exists_add {g : G} (x : add_monoid_algebra k G) {g' : G}
+  (h : ∃ d, g' = g + d):
+  mod_of g x g' = 0 :=
+finsupp.filter_apply_neg _ _ $ by rwa [not_not]
+
+@[simp] lemma mod_of_apply_add {g : G} (x : add_monoid_algebra k G) (d : G) :
+  mod_of g x (d + g) = 0 :=
+mod_of_apply_of_exists_add _ ⟨_, add_comm _ _⟩
+
+@[simp] lemma mod_of_apply_add' {g : G} (x : add_monoid_algebra k G) (d : G) :
+  mod_of g x (g + d) = 0 :=
+mod_of_apply_of_exists_add _ ⟨_, rfl⟩
+
+lemma single_mul_apply_of_not_exists_add (r : k) {g g' : G} (x : add_monoid_algebra k G)
+  (h : ¬∃ d, g' = g + d):
+  (finsupp.single g r * x : add_monoid_algebra k G) g' = 0 :=
+begin
+  classical,
+  rw [mul_apply, finsupp.sum_single_index],
+  swap,
+  { simp_rw [finsupp.sum, zero_mul, if_t_t, finset.sum_const_zero] },
+  { apply finset.sum_eq_zero,
+    simp_rw ite_eq_right_iff,
+    rintros g'' hg'' rfl,
+    exfalso,
+    exact h ⟨_, rfl⟩ },
+end
+
+lemma mul_single_apply_of_not_exists_add (r : k) {g g' : G} (x : add_monoid_algebra k G)
+  (h : ¬∃ d, g' = g + d):
+  (x * finsupp.single g r : add_monoid_algebra k G) g' = 0 :=
+begin
+  classical,
+  rw [mul_apply, finsupp.sum_comm, finsupp.sum_single_index],
+  swap,
+  { simp_rw [finsupp.sum, mul_zero, if_t_t, finset.sum_const_zero] },
+  { apply finset.sum_eq_zero,
+    simp_rw ite_eq_right_iff,
+    rintros g'' hg'' rfl,
+    exfalso,
+    exact h ⟨_, add_comm _ _⟩ }
+end
+
+lemma mod_of_of'_mul (g : G) (x : add_monoid_algebra k G) :
+  mod_of g (of' k G g * x) = 0 :=
+begin
+  ext g',
+  rw finsupp.zero_apply,
+  obtain ⟨d, rfl⟩ | h := em (∃ d, g' = g + d),
+  { rw mod_of_apply_add' },
+  { rw [mod_of_apply_of_not_exists_add _ h, of'_apply, single_mul_apply_of_not_exists_add _ _ h] },
```

Comment:
Not quite in scope but I'd like your opinions on the following: I've seen this occur sometimes where there's a pair of results like `P x y → foo x y = bar` and `¬(P x y) → foo x y = baz`, and it's always slightly inconvenient to do the case distinction. Either you need `em`/`by_cases` and rewrite each case, or to rewrite `foo x y = if P x y then bar else baz` and then `split_ifs`. Is there no way to do this in a single step?